Handle Pandas time indices within StationPlots

zbruick opened the pull request:
### Description Of Changes
Adds support for Pandas DatetimeIndex and TimedeltaIndex as valid `x` variables in `StationPlot`. This addresses #957, which shows a use case for a meteogram of sky cover (and weather symbols could be imagined too). If needed, other datetime types could be added for further support.

This also adds a dependency on Pandas, which I think we're now up to 3 PRs in the 0.11 queue with this dependency.
#### Checklist
<!--
Feel free to remove check-list items aren't relevant to your change

Please use keywords (e.g., Fixes, Closes) to create link to the issues or pull
requests you resolved, so that they will automatically be closed when your pull
request is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
- [x] Closes #957 
- [x] Tests added

dopplershift commented on metpy/plots/station_plot.py:
```diff
@@ -59,6 +61,10 @@ def __init__(self, ax, x, y, fontsize=10, spacing=None, transform=None, **kwargs
 
         """
+        # Handle datetime indices as independent variables
+        if isinstance(x, (pd.DatetimeIndex, pd.TimedeltaIndex)):
```
This is ok if we *have* to handle it here. I'd like to handle it more in a matplotlib-like way. Where does the error come out of if we don't do this?

zbruick replied:
It's a "TypeError: invalid type promotion":

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-8de7bfcbd1f2> in <module>
     31 #x= dates.date2num(ts.index)   # works!  ... Thank you Ryan!
     32 stationplot = StationPlot(ax, x , y, fontsize=12)
---> 33 stationplot.plot_symbol('C', ts , sky_cover)
     34 
     35 ax.xaxis.set_major_locator(mdates.DayLocator())

~/git_repos/MetPy/metpy/plots/station_plot.py in plot_symbol(self, location, codes, symbol_mapper, **kwargs)
    162         # Make sure we use our font for symbols
    163         kwargs['fontproperties'] = wx_symbol_font.copy()
--> 164         return self.plot_parameter(location, codes, symbol_mapper, **kwargs)
    165 
    166     def plot_parameter(self, location, parameter, formatter='.0f', **kwargs):

~/git_repos/MetPy/metpy/plots/station_plot.py in plot_parameter(self, location, parameter, formatter, **kwargs)
    198         """
    199         text = self._to_string_list(parameter, formatter)
--> 200         return self.plot_text(location, text, **kwargs)
    201 
    202     def plot_text(self, location, text, **kwargs):

~/git_repos/MetPy/metpy/plots/station_plot.py in plot_text(self, location, text, **kwargs)
    233         text_collection = self.ax.scattertext(self.x, self.y, text, loc=location,
    234                                               size=kwargs.pop('fontsize', self.fontsize),
--> 235                                               **kwargs)
    236         if location in self.items:
    237             self.items[location].remove()

~/git_repos/MetPy/metpy/plots/_mpl.py in scattertext(self, x, y, texts, loc, **kw)
    213         # Add it to the axes and update range
    214         self.add_artist(text_obj)
--> 215         self.update_datalim(text_obj.get_datalim(self.transData))
    216         self.autoscale_view()
    217         return text_obj

~/git_repos/MetPy/metpy/plots/_mpl.py in get_datalim(self, transData)
    274             """
    275             full_transform = self.get_transform() - transData
--> 276             XY = full_transform.transform(np.vstack((self.x, self.y)).T)  # noqa: N806
    277             bbox = transforms.Bbox.null()
    278             bbox.update_from_data_xy(XY, ignore=True)

~/miniconda3/envs/devel/lib/python3.7/site-packages/numpy/core/shape_base.py in vstack(tup)
    281     """
    282     _warn_for_nonsequence(tup)
--> 283     return _nx.concatenate([atleast_2d(_m) for _m in tup], 0)
    284 
    285 

TypeError: invalid type promotion
```

dopplershift replied:
Hrmm...need to look at matplotlib's unit-handling code for `ax.text()` or `Text`. I don't think they need anything special to handle these Pandas classes, but we might need to borrow some updated code for our `TextCollection`.

zbruick replied:
Got it. Just copied some of the unit conversion code from the `draw` method.

dopplershift commented on setup.py:
```diff
@@ -53,7 +53,7 @@
                       'pint', 'xarray>=0.10.7', 'enum34;python_version<"3.4"',
                       'contextlib2;python_version<"3.6"',
-                      'pooch>=0.1', 'traitlets>=4.3.0'],
+                      'pooch>=0.1', 'traitlets>=4.3.0', 'pandas'],
```
Probably should figure out a version to support (whatever was released ~2 years ago) and update the Travis config to include this so we make sure we actually support our "minimum pandas".

zbruick replied:
Looks like 0.21.0 is closest to two years ago. I'll add 0.24.2 to the 2.7 config for Travis, since that's the last release that supports 2.7.

dopplershift replied:
I think we should use 0.21.0 (or even 0.20.0) as the minimum. That goes back to May 2017.

zbruick commented:
Got the unit handling figured out I think. I added an image test (based on #957), but since this is my first to do, please let me know if I need to change anything. I dropped all of the tolerances down to 0 and it passed on my machine, so I'm curious how CI does with it, or if the test isn't doing what I think it is.

dopplershift commented:
Do we need to add the Pandas dependency here if we're not actually using it?

dopplershift commented:
Oh, it's used in testing...

zbruick commented:
We already have it as a testing dependency I believe (#1034). So I think it could be removed from this PR.

jthielen commented:
I might be missing something, but isn't pandas already a dependency of MetPy, since pandas is a hard dependency of xarray? It would just be determining which version to test?

dopplershift commented on metpy/plots/tests/test_station_plot.py:
```diff
@@ -341,3 +342,21 @@ def test_barb_unit_conversion_exception(u, v):
         stnplot.plot_barb(u, v, plot_units='knots')
+
+
+@pytest.mark.mpl_image_compare(tolerance={'1.5': 0.0, '1.4': 0.0}.get(MPL_VERSION, 0.000),
```
Given that we only support matplotlib >= 2.0, you should be able to just make this `tolerance=0`.

dopplershift replied:
Given that we only support matplotlib >= 2.0, you should be able to just make this tolerance=0.

dopplershift commented:
@zbruick You might cherry-pick the commits onto another of your PRs to save the work. We might ship 0.11 without the other PRs  (who knows) so to me it seems better to only add it as a strict dependency when we actually have it.

@jthielen While the environment will have Pandas because of Xarray, we should only be listing dependencies on things that MetPy actually imports in the library proper. That way if for some reason xarray drops its dependence on Pandas, we're not forcing it needlessly in our packaging. Probably not a practical difference here, but a good policy to follow.

Likewise, we should have a listing for everything we import and not just rely on it being pulled in transitively from something else we depend on.

zbruick commented:
Yeah I agree with the comments here. No reason to add it in this PR - I forgot I was dropping what I was using it for with the update. Good suggestion to cherrypick them over before dropping them. Fine with me to ship 0.11 without the DSG PR, as I'm guessing the appending stuff I added may take some time to review and revise.

zbruick commented:
Pandas dependency has been removed, and this PR has been rebased.
